src/runtime/virtcontainers/kata_agent.go

```diff
@@ -1568,6 +1577,63 @@ func (k *kataAgent) handleDeviceBlockVolume(c *Container, m Mount, device api.De
 	default:
 		return nil, fmt.Errorf("Unknown block device driver: %s", c.sandbox.config.HypervisorConfig.BlockDeviceDriver)
 	}
+	return vol, nil
+}
+
+func handleImageGuestPullBlockVolume(c *Container, virtualVolumeInfo *types.KataVirtualVolume, vol *grpc.Storage) (*grpc.Storage, error) {
+	container_annotations := c.GetAnnotations()
+	container_type := container_annotations["io.kubernetes.cri.container-type"]
+	if virtualVolumeInfo.Source == "" {
```

Comment:
I am confused by this test: if we reach this function, it means we have received a kata virtual volume configured for guest pull image. In what situation can it have an empty "Source"? Shouldn't we see it as an error, rather than trying to guess the missing information?
Put another way: if we can get the image name from annotations, why do we need the "Source" field in the kata virtual volume structure?

Reply:
Thanks @littlejawa . I have fixed it. In guest pull, we have the option to disable this field for now.. 

Besides pulling guest images, this field also supports raw block, direct volume and nydus image. However, containerd does not support to pass sandbox images into annotations yet (https://github.com/containerd/containerd/pull/9419). Once this issue is resolved, we will be able to access the image name from the source field directly without parsing the annotations.

Reply:
This actually clarifies some questions I had when working on supporting it on the cri-o side (see https://github.com/cri-o/cri-o/pull/7471).
I've used the "Source" field in cri-o to provide the image name, but I'm also giving the annotations, so with the added cri-o annotation support, this will work fine.
Thanks!